source/extensions/filters/http/adaptive_concurrency/controller/gradient_controller.cc

```diff
@@ -215,10 +215,12 @@ void GradientController::recordLatencySample(MonotonicTime rq_send_time) {
     sample_count = hist_sample_count(latency_sample_hist_.get());
   }
 
-  if (inMinRTTSamplingWindow() && sample_count >= config_.minRTTAggregateRequestCount()) {
+  const auto agg_count = config_.minRTTAggregateRequestCount();
+  if (min_rtt_update_mtx_.TryLock() && inMinRTTSamplingWindow() && sample_count >= agg_count) {
     // This sample has pushed the request count over the request count requirement for the minRTT
     // recalculation. It must now be finished.
     updateMinRTT();
+    min_rtt_update_mtx_.Unlock();
```

Comment:
It would be good to have a scoped try-lock primitive to avoid leaking the lock due to a logic error.

Reply:
I don't think we can have a scoped try-lock object since the absl scoped mutex primitives guarantee that they hold the lock if the objects are constructed. I know Facebook's Folly library has something that allows for this with pointers, but didn't see anything in Mutex.h.

Take a look at the new version of this patch and let me know if you think it looks less error-prone.